src/lib.rs

```diff
@@ -160,67 +184,135 @@ pub trait Get<T>: Collection {
 /// Mutably queryable collection.
 pub trait GetMut<T>: Get<T> {
 	/// Returns a mutable reference to the item stored behind the given key (if any).
-	fn get_mut(&mut self, key: T) -> Option<&mut Self::Item>;
+	fn get_mut<'a>(&'a mut self, key: T) -> Option<Self::ItemMut<'a>>;
+}
+
+/// Guarded queryable collection.
+pub trait GetLock<T>: Get<T> {
+	/// Returns a guarded reference to the item stored behind the given key (if any).
+	///
+	/// It is expected that this method may block.
+	fn get_lock<'a>(&'a self, key: T) -> Option<Self::ItemGuard<'a>>;
 }
 
 /// Collection exposing a reference to its front element.
 pub trait Front: Collection {
 	/// Get a reference to the front element of the collection.
-	fn front(&self) -> Option<&Self::Item>;
+	fn front<'a>(&'a self) -> Option<Self::ItemRef<'a>>;
 }
 
 /// Collection exposing a reference to its back element.
 pub trait Back: Collection {
 	/// Get a reference to the back element of the collection.
-	fn back(&self) -> Option<&Self::Item>;
+	fn back<'a>(&'a self) -> Option<Self::ItemRef<'a>>;
 }
 
 /// Collection exposing a mutable reference to its front element.
 pub trait FrontMut: Collection {
 	/// Get a mutable reference to the front element of the collection.
-	fn front_mut(&mut self) -> Option<&mut Self::Item>;
+	fn front_mut<'a>(&'a mut self) -> Option<Self::ItemMut<'a>>;
 }
 
 /// Collection exposing a mutable reference to its back element.
 pub trait BackMut: Collection {
 	/// Get a mutable reference to the back element of the collection.
-	fn back_mut(&mut self) -> Option<&mut Self::Item>;
+	fn back_mut<'a>(&'a mut self) -> Option<Self::ItemMut<'a>>;
+}
+
+/// Collection exposing a guarded reference to its front element.
+pub trait FrontLock: Collection {
+	/// Get a guarded referencce to the front element of the collection.
+	///
+	/// It is expected that this method may block.
+	fn front_lock<'a>(&'a self) -> Option<Self::ItemGuard<'a>>;
+}
+
+/// Collection exposing a guarded reference to its back element.
+pub trait BackLock: Collection {
+	/// Get a guarded reference to the back element of the collection.
+	///
+	/// It is expected that this method may block.
+	fn back_lock<'a>(&'a self) -> Option<Self::ItemGuard<'a>>;
 }
 
 /// Mutable collection where new elements can be inserted.
 pub trait Insert: Collection {
 	/// The output of the insertion function.
-	type Output;
+	type Output<'a>;
```

Comment:
Do you have a motivation for adding a lifetime to the associated `Output` types?